llvm/lib/ABI/Targets/X86.cpp

```diff
@@ -0,0 +1,523 @@
+    if (ST->hasFlexibleArrayMember()) {
+      return;
+    }
+    // Reset Lo class, this will be recomputed.
+    Current = NoClass;
+
+    // If this is a C++ record, classify the bases first.
+    if (ST->isCXXRecord()) {
+      const FieldInfo *BaseClasses = ST->getBaseClasses();
+      for (uint32_t I = 0; I < ST->getNumBaseClasses(); ++I) {
+        const FieldInfo &Base = BaseClasses[I];
+
+        // Classify this field.
+        //
+        // AMD64-ABI 3.2.3p2: Rule 3. If the size of the aggregate exceeds a
+        // single eightbyte, each is classified separately. Each eightbyte gets
+        // initialized to class NO_CLASS.
+        Class FieldLo, FieldHi;
+        uint64_t Offset = OffsetBase + Base.OffsetInBits;
+        classify(Base.FieldType, Offset, FieldLo, FieldHi, IsNamedArg);
+        Lo = merge(Lo, FieldLo);
+        Hi = merge(Hi, FieldHi);
+
+        if (getABICompatInfo().Flags.ReturnCXXRecordGreaterThan128InMem &&
+            (Size > 128 &&
+             (Size != Base.FieldType->getSizeInBits().getFixedValue() ||
+              Size > getNativeVectorSizeForAVXABI(AVXLevel)))) {
+          Lo = Memory;
+          postMerge(Size, Lo, Hi);
+          return;
+        }
+
+        if (Lo == Memory || Hi == Memory) {
+          postMerge(Size, Lo, Hi);
+          return;
+        }
+      }
+    }
+
+    // Classify the fields one at a time, merging the results.
+    const FieldInfo *Fields = ST->getFields();
+    uint32_t NumFields = ST->getNumFields();
+
+    for (uint32_t I = 0; I < NumFields; ++I) {
+      const FieldInfo &Field = Fields[I];
+      uint64_t Offset = OffsetBase + Field.OffsetInBits;
+      bool BitField = Field.IsBitField;
+
+      if (Size > 128 &&
+          Size != Field.FieldType->getSizeInBits().getFixedValue() &&
+          Size > getNativeVectorSizeForAVXABI(AVXLevel)) {
+        Lo = Memory;
+        postMerge(Size, Lo, Hi);
+        return;
+      }
+      if (!BitField) {
+        uint64_t FieldAlign = Field.FieldType->getAlignment().value() * 8;
+        if (Offset % FieldAlign) {
+          Lo = Memory;
+          postMerge(Size, Lo, Hi);
+          return;
+        }
+      }
+
+      Class FieldLo, FieldHi;
+
+      if (BitField) {
+        uint64_t BitFieldSize = Field.BitFieldWidth;
+        uint64_t EB_Lo = Offset / 64;
+        uint64_t EB_Hi = (Offset + BitFieldSize - 1) / 64;
+
+        if (EB_Lo) {
+          assert(EB_Hi == EB_Lo && "Invalid classification, type > 16 bytes.");
+          FieldLo = NoClass;
+          FieldHi = Integer;
+        } else {
+          FieldLo = Integer;
+          FieldHi = EB_Hi ? Integer : NoClass;
+        }
+      } else {
+        classify(Field.FieldType, Offset, FieldLo, FieldHi, IsNamedArg);
+      }
+
+      Lo = merge(Lo, FieldLo);
+      Hi = merge(Hi, FieldHi);
+      if (Lo == Memory || Hi == Memory)
+        break;
+    }
+    postMerge(Size, Lo, Hi);
+    return;
+  }
+  if (const auto *UT = dyn_cast<UnionType>(T)) {
```

Comment:
We're missing handling for UseClang11Compat here.

Another notes is that clang handles structs + unions with common code, because both are "records". Not entirely sure which was is better.